File: backend/app/schemas/order.py
from pydantic import BaseModel
from app.models.order import OrderStatus
from typing import List
from decimal import Decimal

ALLOWED_TRANSITIONS = {
    OrderStatus.OPEN: [OrderStatus.SENT, OrderStatus.CANCELLED],
    OrderStatus.SENT: [OrderStatus.IN_PROGRESS],
    OrderStatus.IN_PROGRESS: [OrderStatus.READY],
    OrderStatus.READY: [OrderStatus.CLOSED],
}

class OrderItemOut(BaseModel):
    id: int
    product_name: str
    quantity: int
    unit_price: float
    subtotal: float
    status: str
#class OrderItemOut(BaseModel):
#    product_name: str
#    quantity: int
#    unit_price: Decimal

class OrderOut(BaseModel):
    order_id: int
    table_number: int
    status: str
    items: List[OrderItemOut]
    total: Decimal

class OrderStatusUpdate(BaseModel):
    status: OrderStatus

File: frontend/src/types/order.ts
import type { OrderStatus } from "./orderStatus"
import type { OrderItemStatus } from "./orderItemStatus"
import type { PaymentMethod } from "./paymentMethod"

export interface OrderItem {
    id: number
    product_name: string
    quantity: number
    status: OrderItemStatus
}

export interface Order {
    id: number
    table_number: number
    status: OrderStatus
    created_at: string
    items: OrderItem[]
}

export interface OrderDetailItem {
  id: number
  product_name: string
  quantity: number
  unit_price: number
  status: OrderItemStatus
  notes?: string | null
}


export interface OrderPayment {
  id: number
  amount: number
  method: PaymentMethod
}


export interface OrderDetail {
  id: number
  table_id: number
  table_number: number
  status: OrderStatus
  items: OrderDetailItem[]
  subtotal: number
  payments: OrderPayment[]
  total: number
  total_paid: number
  remaining: number
  discount: number
}

export type RawOrderDetail = Omit<
  OrderDetail,
  | "items"
  | "payments"
  | "subtotal"
  | "total"
  | "total_paid"
  | "remaining"
  | "discount"
> & {
  items: Array<
    Omit<OrderDetailItem, "unit_price"> & {
      unit_price: unknown
    }
  >
  payments: Array<
    Omit<OrderPayment, "amount"> & {
      amount: unknown
    }
  >
  subtotal: unknown
  total: unknown
  total_paid: unknown
  remaining: unknown
  discount: unknown
}

export interface AddProductToTableResponse {
  order_id: number
}

export interface CashierOrder {
  id: number
  table_number: number
  status: OrderStatus
  subtotal: number
  total: number
  total_paid: number
  remaining: number
  discount: number
  payments: OrderPayment[]
}

export type RawOrderPayment =
  Omit<OrderPayment, "amount"> & {
    amount: unknown
  }

export type RawCashierOrder =
  Omit<
    CashierOrder,
    | "subtotal"
    | "total"
    | "total_paid"
    | "remaining"
    | "discount"
    | "payments"
  > & {
    subtotal: unknown
    total: unknown
    total_paid: unknown
    remaining: unknown
    discount: unknown
    payments: RawOrderPayment[]
  }